proposals/NNNN-inheritance-of-actor-isolation.md

```diff
@@ -0,0 +1,551 @@
+
+    // not allowed
+    await counter.incrementAndSleep(isolation: nil)
+  }
+}
+
+@MainActor func testMainActor(counter: Counter) {
+  // allowed
+  await counter.incrementAndSleep(isolation: MainActor.shared)
+
+  // not allowed
+  await counter.incrementAndSleep(isolation: nil)
+}
+
+func testNonIsolated(counter: Counter) {
+  // allowed
+  await counter.incrementAndSleep(isolation: nil)
+
+  // not allowed
+  await counter.incrementAndSleep(isolation: MainActor.shared)
+}
+```
+
+### `#isolation` default argument
+
+The special expression form `#isolation` can be used as a default
+argument:
+
+```swift
+extension Collection {
+  func sequentialMap<R>(isolation: isolated (any Actor)? = #isolation,
+                        transform: (Element) async -> R) async -> [R] {
+    var results: [R] = []
+    for elt in self {
+      results.append(await transform(elt))
+    }
+    return results
+  }
+}
+```
+
+When a call uses this default argument, it behaves as if the argument
+was an expression representing the static actor isolation of the
+current context:
+
+- if the current context is statically non-isolated, the parameter
+  must have optional type, and the argument is `nil`;
+- if the current context is isolated to a global actor `T`, the argument
+  is `T.shared`;
+- if the current context has an `isolated` parameter (including the
+  implicit `self` parameter of an actor method), the argument is a
+  reference to that parameter;
+- otherwise, the current context must be a closure which captures
+  an `isolated` parameter or a non-optional binding of it, and the
+  argument is a reference to that capture.
+
+Except where noted above, a parameter using `#isolation` as a default
+argument can have any type.  When type-checking considers a candidate
+function for a call that would use this default argument for a parameter,
+it assumes that the notional argument expression above can be coerced
+to the parameter type.  If the call is actually resolved to use that
+candidate, the coercion must succeed or the call is ill-formed.
+This rule is necessary in order to avoid the need to decide the isolation
+of the calling context before resolving calls from it.
+
+The parameter does not have to be an `isolated` parameter.
+
+## Source compatibility
+
+This proposal is largely additive and should not affect the behavior
+of existing code.
+
+The new rules for isolation checking permit more calls to be
+recognized as sharing isolation.  This should strictly allow
+more code to be compiled; it cannot cause source-compatibility
+regressions by allowing different overloads to be picked because
+isolation checking is performed separately from type-checking.
+
+## ABI compatibility
+
+This proposal does not change how any existing code is compiled.
+
+## Implications for adoption
+
+This proposal does not add any new types and does not require new
+runtime or library support.  It can be implemented purely in the compiler.
```

Comment:
As noted in the PR for `#isolation`, when implemented as an expression macro, it would tie this feature (and hence, any async features that depend on it) to platform support for macros. Ought we to consider platform bring-up implications here?